src/Http/Controllers/MetricsController.php

```diff
@@ -14,79 +14,73 @@
 
 class MetricsController extends Controller
 {
-
+    /**
+     * @throws Exception
+     */
     public function getBadgeData($start, $end): array
     {
-        $tasks = Task::with('badge')
-            ->whereDate('created_at', '>=', new DateTime($start))
+        $tasks = Task::whereDate('created_at', '>=', new DateTime($start))
             ->whereDate('created_at', '<=', new DateTime($end))
             ->get();
 
-
-        $full = [];
+        $badgeData = [];
         foreach ($tasks as $task) {
-            if (array_key_exists($task->badge->name, $full)) {
```

Comment:
Consider groupby 'name' and count() instead of a foreach loop

Reply:
That's a really good idea. I did it for two of the functions and ill look into refactoring the rest when I move them into actions